content/pda.md

```diff
@@ -0,0 +1,523 @@
+
+```rust
+pub fn try_find_program_address(seeds: &[&[u8]], program_id: &Pubkey) -> Option<(Pubkey, u8)> {
+
+    let mut bump_seed = [std::u8::MAX];
+    for _ in 0..std::u8::MAX {
+        {
+            let mut seeds_with_bump = seeds.to_vec();
+            seeds_with_bump.push(&bump_seed);
+            match Self::create_program_address(&seeds_with_bump, program_id) {
+                Ok(address) => return Some((address, bump_seed[0])),
+                Err(PubkeyError::InvalidSeeds) => (),
+                _ => break,
+            }
+        }
+        bump_seed[0] -= 1;
+    }
+    None
+
+}
+```
+
+The `create_program_address` function performs a set of hash operations over the seeds and `program_id` to compute a key and then verifies if the computed key lies on the ed25519 elliptic curve or not. If a valid PDA is found, then the PDA is returned. Otherwise, an error is returned.
+
+```rust
+pub fn create_program_address(
+        seeds: &[&[u8]],
+        program_id: &Pubkey,
+    ) -> Result<Pubkey, PubkeyError> {
+
+        let mut hasher = crate::hash::Hasher::default();
+        for seed in seeds.iter() {
+            hasher.hash(seed);
+        }
+        hasher.hashv(&[program_id.as_ref(), PDA_MARKER]);
+        let hash = hasher.result();
+
+        if bytes_are_curve_point(hash) {
+            return Err(PubkeyError::InvalidSeeds);
+        }
+
+        Ok(Pubkey::new(hash.as_ref()))
+
+    }
+```
+
+In summary, the `find_program_address` function passes our input seeds and `program_id` to the `try_find_program_address` function. The `try_find_program_address` function adds a `bump_seed` (starting from 255) to our input seeds calls the `create_program_address` function in a loop until a valid PDA is found. Once a valid PDA is found, both the PDA and `bump_seed` are returned.
+
+Note that it is possible for the same input seeds and different valid bumps to generate different valid PDAs. The `bump_seed` returned by `find_program_address` will always be the first valid PDA found, and thus the `bump_seed` with the largest value. This `bump_seed` is commonly referred to as the "canonical bump". It's recommended to only use the canonical bump to avoid confusion and alway validate the PDAs passed into your program.
+
+### Why do PDAs matter?
+
+PDAs are important because it allows us to easily map our program’s accounts. Instead of keeping track of each address, we simply need to remember the seeds used to derive PDAs to find an account.
+
+For example, in programs that store user-specific data it’s common to use a user’s public key as the seed. This separates each user’s data into its own PDA and makes it possible for the client to locate each user’s data by finding the address using the program ID and the user’s public key. One limitation of this approach is that it would limit each user to only one PDA account for our program.
```

Comment:
@ZYJLiu It would be great if this paragraph could exist without using "use" as verb - there's some many "users" 😛

Reply:
gon funnit - sorry, I meant this as a comment not a review